### PR TITLE
✨ Require visibility trigger selector's to have data-vars-*

### PR DESCRIPTION
**Pull request by micajuine-ho**
Task 3 of #26823

- Filter only elements that have `data-vars-*` attribute within `getAmpElements()`

### CODE REVIEW

**Comment by micajuine-ho** [comment]
@zhouyx  PTAL

File: extensions/amp-analytics/0.1/test/test-events.js [hunk 1]
```diff
@@ -1957,33 +1961,53 @@ describes.realWin('Events', {amp: 1}, (env) => {
           readyPromise = Promise.resolve();
           unlisten = env.sandbox.spy();
           unlisten2 = env.sandbox.spy();
+          unlisten3 = env.sandbox.spy();
           config = {};
 
           eventsSpy = env.sandbox.spy(tracker, 'onEvent_');
 
+          target3 = win.document.createElement('div');
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-events.js:1969]: I don't see the need to introduce the `target3` here? Any reason we want a third element.

Comment (micajuine-ho) [reply extensions/amp-analytics/0.1/test/test-events.js:1969]: I felt that it was important to test/show that array selector was using querySelectorAll here... but I guess we talked about this before and agreed that this functionality is handled by analytics root and so should be tested there...

Removing now

**Comment by micajuine-ho** [comment]
@zhouyx Ready for review

File: extensions/amp-analytics/0.1/analytics-root.js [hunk 2]
```diff
@@ -300,7 +302,7 @@ export class AnalyticsRoot {
         }
         for (let j = 0; j < nodeList.length; j++) {
           if (this.contains(nodeList[j])) {
-            elementArray.push(nodeList[j]);
+            this.checkElementDataVars(nodeList[j], elementArray, selector);
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/analytics-root.js:305]: nit: Can we make this a private method. 
Personally I would expect the `checkElementDataVars` to return a boolean based on the method name, and you won't need to pass the `elementArray` or `selector`. Not blocking.

File: extensions/amp-analytics/0.1/analytics-root.js [hunk 3]
```diff
@@ -314,6 +316,33 @@ export class AnalyticsRoot {
...
+      getDataParamsFromAttributes(
+        element,
+        /* computeParamNameFunc */ undefined,
+        VARIABLE_DATA_ATTRIBUTE_KEY
+      )
+    );
+    if (dataVarKeys.length) {
+      elementArray.push(element);
+    } else {
+      user().warn(
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/analytics-root.js:337]: When multiple elements are selected. I think the message here will get duplicated. 

One option is to also print the element information. A better option is to only print the warning message once with the number of elements that don't have `data-vars-` attribute specified. WDYT?

Comment (micajuine-ho) [reply extensions/amp-analytics/0.1/analytics-root.js:344]: Yes I agree, changed up the method to do slightly more work.

File: extensions/amp-analytics/0.1/analytics-root.js (hunk 2 above)
Comment (micajuine-ho) [reply extensions/amp-analytics/0.1/analytics-root.js]: Changed the method.

File: extensions/amp-analytics/0.1/test/test-analytics-root.js [hunk 4]
```diff
@@ -739,23 +763,38 @@ describes.realWin(
...
+        const child4 = win.document.createElement('child');
         // Parent child attached to parent doc should not be captured
         const parentChild = env.parentWin.document.createElement('child');
         body.appendChild(child2);
         body.appendChild(child3);
+        body.appendChild(child4);
         env.parentWin.document.body.appendChild(parentChild);
         child.classList.add('myClass');
         child2.classList.add('myClass');
+        child4.classList.add('myClass');
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/test/test-analytics-root.js:778]: can we remove this? Why we need child4?

Comment (micajuine-ho) [reply extensions/amp-analytics/0.1/test/test-analytics-root.js:778]: Child4 has no data-vars-* attribute, so it's testing that it doesn't get picked up (I'll add a comment).

Comment (zhouyx) [reply extensions/amp-analytics/0.1/test/test-analytics-root.js:779]: My bad! Complete missed that part. 

I understand this is the test case for embed and it's duplicate version of the test above. But this unit test covers
1.selector select all elements
2. select element within the doc
3. data-vars

Maybe consider breaking it to multiple unit tests next time.

Comment (micajuine-ho) [reply extensions/amp-analytics/0.1/test/test-analytics-root.js]: Done :)

**Pull request merged**